redisuniversal/redis.go

```diff
@@ -0,0 +1,229 @@
+package redisuniversal
+
+import (
+"encoding/json"
+"strings"
+"fmt"
+"os"
+"os/signal"
+"time"
+
+"github.com/go-redis/redis"
+"gopkg.in/launchdarkly/go-server-sdk.v4"
+"gopkg.in/launchdarkly/go-server-sdk.v4/ldlog"
+"gopkg.in/launchdarkly/go-server-sdk.v4/utils"
+)
+
+type redisStoreCore struct {
+	loggers ldlog.Loggers
+	config  Options
+	client  redis.UniversalClient
+	clusterMode bool
+}
+
+const hashTag = "{ld}."
+
+func newRedisStoreCore(options Options, loggers ldlog.Loggers) *redisStoreCore {
+	loggers.SetPrefix("RedisFeatureStore:")
+	client := redis.NewUniversalClient(options.RedisOpts)
+
+	// ping the server so we know we are good
+	err := client.Ping().Err()
+	if err != nil {
+		loggers.Error("could not ping redis server: %v", err)
+		return nil
+	}
+
+	c := make(chan os.Signal)
+	signal.Notify(c, os.Interrupt, os.Kill)
+	go func() {
+		select {
+		case <-c:
```

Comment:
Due to the select, this is a non-blocking read, which I believe means that this goroutine is going to start, find that there's nothing on the channel, and immediately stop. I think the intention was that it should wait until `signal.Notify` pushes a value.

However, I'm a little uncomfortable with this use of `signal.Notify`. It has a global effect on the process, i.e. the normal behavior of these signals is turned off, so it's not the kind of thing a library would normally do. Could you say more about why it's important to close the Redis client before the process exits for any reason? Normally, exiting a process automatically closes all open sockets, at least in Unix-based OSes. And the go-redis documentation, as far as I can see, doesn't say that it is crucial to call Close in this situation; in fact it says most applications will never need to call Close.

Reply:
Also, `os.Kill` is not a signal that can be caught.